### PR TITLE
:rocket: feat: add error handling chapter

### LINKED ISSUE #79: Recoverable Errors with `Result<T,E>`
- Explain how you can handle an execution error by introducing the `Result` enum
- Give an example with existing core library operations (handling e.g. overflowing addition)
- Shortcuts to panic on error: `expect`, `unwrap` methods defined on the `ResultTrait` trait
- Error propagation (see [rustbook](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#propagating-errors))
- The `?` operator

As this concept shares a lot of similarities with Rust, the submitted PR will follow the content of the Rust book, adapting all examples to Cairo, and removing concepts that don't have an equivalent. The examples will need to make sense in the context in the execution of a Cairo program.

Estimated time: 2-3 hours

### PR DESCRIPTION
**Pull request by 0xEniotna**
close #79

### CODE REVIEW

File: src/ch09-01-error-handling.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+---
+
+<br />
+
+Most errors aren’t serious enough to require the program to stop entirely. Sometimes, when a function fails, it’s for a reason that you can easily interpret and respond to. For example, if you try to open a file and that operation fails because the file doesn’t exist, you might want to create the file instead of terminating the process.
+
+## The `Result` enum
+
+Recall from [“Generic data types”](/src/ch07-01-generic-data-types.md#enums) in Chapter 7 that the `Result` enum is defined as having two variants, `Ok` and `Err`, as follows:
```
Comment (enitrat) [inline src/ch09-01-error-handling.md:11]: Incorrect link
ch07-01-generic-data-types.md#enums

File: src/ch09-01-error-handling.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+}
+```
+
+The `expect` and `unwrap` methods are similar in that they both attempt to extract the value of type `T` from a `Result<T, E>` when it is in the `Ok` variant. If the `Result` is `Ok(x)`, both methods return the value `x`. However, the key difference between the two methods lies in their behavior when the `Result` is in the `Err` variant. The `expect` method allows you to provide a custom error message (as a `felt252` value) that will be used when panicking, giving you more control and context over the panic. On the other hand, the `unwrap` method panics with a default error message, providing less information about the cause of the panic.
+
+The `expect_err` and `unwrap_err` have to exact opposite behavior. If the `Result` is `Err(x)`, both methods return the value `x`. However, the key difference between the two methods is in case of `Result::Ok()`. The `expect_err` method allows you to provide a custom error message (as a `felt252` value) that will be used when panicking, giving you more control and context over the panic. On the other hand, the `unwrap_err` method panics with a default error message, providing less information about the cause of the panic.
```
Comment (enitrat) [inline src/ch09-01-error-handling.md:44]: > have to exact opposite behavior.

File: src/ch09-01-error-handling.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+A careful reader may have noticed the `<impl TDrop: Drop<T>>` and `<impl EDrop: Drop<E>>` in the first four methods signatures. This syntax represents generic type constraints in the Cairo language. These constraints indicate that the associated functions require an implementation of the `Drop` trait for the generic types `T` and `E`, respectively. The `Drop` trait is used to define custom clean-up logic for values when they go out of scope, ensuring proper resource management. By specifying these constraints, the functions using them guarantee that the values of types `T` and `E` will be cleaned up correctly, whether they are part of the `Ok` or `Err` variant of the `Result` enum.
```
Comment (enitrat) [inline src/ch09-01-error-handling.md:46]: > The `Drop` trait is used to define custom clean-up logic for values when they go out of scope, ensuring proper resource management. By specifying these constraints, the functions using them guarantee that the values of types `T` and `E` will be cleaned up correctly, whether they are part of the `Ok` or `Err` variant of the `Result` enum.

remove this part

File: src/ch09-01-error-handling.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+}
+```
+
+The console will print the error "Invalid Integer".
+
+<br/>
+
+### Summary
+
+That is it for the error handling system in Cairo. We saw that recoverable errors can be handled in Cairo using the Result enum, which has two variants: `Ok` and `Err`. The `Result<T, E>` enum is generic, with types `T` and `E` representing the successful and error values, respectively. The `ResultTrait` provides methods for working with `Result<T, E>`, such as unwrapping values, checking if the result is `Ok` or `Err`, and panicking with custom messages.
```
Comment (enitrat) [inline src/ch09-01-error-handling.md:166]: > That is it for the error handling system in Cairo.

remove this sentence

File: src/ch09-01-error-handling.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+To handle recoverable errors, a function can return a `Result` type and use pattern matching to handle the success or failure of an operation. The `?` operator can be used to implicitly handle errors by propagating the error or unwrapping the successful value. This allows for more concise and idiomatic error handling, where calling functions deal with errors.
```
Comment (enitrat) [inline src/ch09-01-error-handling.md:168]: > where calling functions deal with errors.

not very clear, can you rephrase?

**Review by enitrat — COMMENTED** [review]
good PR, some stuff to address 🥳

**Comment by 0xEniotna** [comment]
Good catches, fixed and pushed

**Comment by fricoben** [comment]
can you edit your first message and add the issue you are focusing on with `close #78` please (for tracking purposes).

**Review by enitrat — APPROVED** [review]
LGTM
